client/keystore/src/lib.rs

```diff
@@ -86,410 +87,132 @@ impl std::error::Error for Error {
-	///
-	/// Does not place it into the file system store.
-	fn insert_ephemeral_pair<Pair: PairT>(&mut self, pair: &Pair, seed: &str, key_type: KeyTypeId) {
-		let key = (key_type, pair.public().to_raw_vec());
-		self.additional.insert(key, seed.into());
-	}
-
-	/// Insert a new key with anonymous crypto.
-	///
-	/// Places it into the file system store.
-	fn insert_unknown(&self, key_type: KeyTypeId, suri: &str, public: &[u8]) -> Result<()> {
-		if let Some(path) = self.key_file_path(public, key_type) {
-			let mut file = File::create(path).map_err(Error::Io)?;
-			serde_json::to_writer(&file, &suri).map_err(Error::Json)?;
-			file.flush().map_err(Error::Io)?;
-		}
-		Ok(())
-	}
-
-	/// Insert a new key.
-	///
-	/// Places it into the file system store.
-	pub fn insert_by_type<Pair: PairT>(&self, key_type: KeyTypeId, suri: &str) -> Result<Pair> {
-		let pair = Pair::from_string(
-			suri,
-			self.password()
-		).map_err(|_| Error::InvalidSeed)?;
-		self.insert_unknown(key_type, suri, pair.public().as_slice())
-			.map_err(|_| Error::Unavailable)?;
-		Ok(pair)
-	}
-
-	/// Insert a new key.
-	///
-	/// Places it into the file system store.
-	pub fn insert<Pair: AppPair>(&self, suri: &str) -> Result<Pair> {
-		self.insert_by_type::<Pair::Generic>(Pair::ID, suri).map(Into::into)
-	}
-
-	/// Generate a new key.
-	///
-	/// Places it into the file system store.
-	pub fn generate_by_type<Pair: PairT>(&self, key_type: KeyTypeId) -> Result<Pair> {
-		let (pair, phrase, _) = Pair::generate_with_phrase(self.password());
-		if let Some(path) = self.key_file_path(pair.public().as_slice(), key_type) {
-			let mut file = File::create(path)?;
-			serde_json::to_writer(&file, &phrase)?;
-			file.flush()?;
-		}
-		Ok(pair)
-	}
-
-	/// Generate a new key.
-	///
-	/// Places it into the file system store.
-	pub fn generate<Pair: AppPair>(&self) -> Result<Pair> {
-		self.generate_by_type::<Pair::Generic>(Pair::ID).map(Into::into)
-	}
-
-	/// Create a new key from seed.
-	///
-	/// Does not place it into the file system store.
-	pub fn insert_ephemeral_from_seed_by_type<Pair: PairT>(
-		&mut self,
-		seed: &str,
-		key_type: KeyTypeId,
-	) -> Result<Pair> {
-		let pair = Pair::from_string(seed, None).map_err(|_| Error::InvalidSeed)?;
-		self.insert_ephemeral_pair(&pair, seed, key_type);
-		Ok(pair)
-	}
-
-	/// Create a new key from seed.
-	///
-	/// Does not place it into the file system store.
-	pub fn insert_ephemeral_from_seed<Pair: AppPair>(&mut self, seed: &str) -> Result<Pair> {
-		self.insert_ephemeral_from_seed_by_type::<Pair::Generic>(seed, Pair::ID).map(Into::into)
-	}
-
-	/// Get the key phrase for a given public key and key type.
-	fn key_phrase_by_type(&self, public: &[u8], key_type: KeyTypeId) -> Result<String> {
-		if let Some(phrase) = self.get_additional_pair(public, key_type) {
-			return Ok(phrase.clone())
-		}
+/// A keystore implementation which uses a backend
+/// that is either local or remote.
+pub struct Keystore(Box<dyn CryptoStore>);
```

Comment:
Could you explain why we need this `Keystore` abstraction? As far as I can tell in its `CryptoStore` implementation all methods defer to `self.0` thus I don't understand the benefit of it.

Reply:
You're right... i started with a specific design, changed it along the way and kept this. I removed this.